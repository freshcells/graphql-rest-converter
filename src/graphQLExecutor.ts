import { execute, parse, GraphQLSchema, ExecutionResult } from 'graphql'
import { GraphQLClient } from 'graphql-request'

export type GraphQLExecutorArgs = {
  document: string
  variables: { [key: string]: unknown }
}

export type GraphQLExecutor = (args: GraphQLExecutorArgs) => Promise<ExecutionResult>

type ExecutorArgs = ConstructorParameters<typeof GraphQLClient>

export const createHttpExecutor = (
  url: ExecutorArgs[0],
  requestConfig: ExecutorArgs[1]
): GraphQLExecutor => {
  const client = new GraphQLClient(url, {
    ...requestConfig,
    errorPolicy: 'all',
  })
  return async ({ document, variables }) => {
    return (await client.rawRequest({
      query: document,
      variables,
    })) as ExecutionResult
  }
}

export const createSchemaExecutor = (schema: GraphQLSchema): GraphQLExecutor => {
  return ({ document, variables }) =>
    Promise.resolve(execute({ schema, document: parse(document), variableValues: variables }))
}
