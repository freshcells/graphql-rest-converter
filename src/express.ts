/* eslint-disable @typescript-eslint/ban-ts-comment */
import express, { RequestHandler, Request, Response } from 'express'
import { OpenAPIV3 } from 'openapi-types'
import { parse, buildSchema } from 'graphql'
import { getBridgeOperations } from './graphql.js'
import { pathTemplateToExpressRoute } from './pathTemplate.js'
import {
  BridgeOperation,
  BridgeOperations,
  CustomOperationProps,
  CreateOpenAPIGraphQLBridgeConfig,
  CreateOpenAPISchemaConfig,
  CreateMiddlewareConfig,
} from './types.js'
import { GraphQLExecutor } from './graphQLExecutor.js'
import { createOpenAPISchemaWithValidate } from './utils.js'
import { createRequestHandler } from './createRequestHandler.js'

const middlewareToPromise =
  (middleware: RequestHandler) =>
  (req: Parameters<RequestHandler>[0], res: Parameters<RequestHandler>[1]) => {
    return new Promise<ReturnType<RequestHandler>>((resolve, reject) => {
      const next = (x: unknown) => (x ? reject(x) : resolve())
      middleware(req, res, next)
    })
  }

const jsonBodyParserPromise = middlewareToPromise(express.json())
const formBodyParserPromise = middlewareToPromise(express.urlencoded({ extended: true }))

const createExpressMiddleware = <
  CustomProps extends CustomOperationProps = CustomOperationProps,
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response,
>(
  operations: BridgeOperations<CustomProps>,
  executor: GraphQLExecutor<ThisRequest, ThisResponse>,
  config?: CreateMiddlewareConfig<ThisRequest, ThisResponse>,
) => {
  const router = express.Router()
  createRequestHandler<ThisRequest, ThisResponse, CustomProps>(
    operations,
    {
      async transformJsonBody(req, res) {
        await jsonBodyParserPromise(req, res)
      },
      async transformFormBody(req, res) {
        await formBodyParserPromise(req, res)
      },
      registerRoute(operation: BridgeOperation<CustomProps>, handleRequest) {
        const route = pathTemplateToExpressRoute(operation.path)
        router[operation.httpMethod](route, async (req, res, next) => {
          try {
            const handlerResult = await handleRequest(req as ThisRequest, res as ThisResponse)
            const { contentType, code, result } = handlerResult
            if (typeof result === 'object') {
              res.status(code).json(result)
              return
            }
            if (contentType) {
              res.contentType(contentType)
            }
            res.status(code).send(result)
          } catch (e) {
            next(e)
          }
        })
      },
    },
    executor,
    config,
  )

  return router
}

export const createOpenAPIGraphQLBridge = <
  ThisRequest extends Request = Request,
  ThisResponse extends Response = Response,
  T extends CustomOperationProps = CustomOperationProps,
>(
  config: CreateOpenAPIGraphQLBridgeConfig<T>,
) => {
  const { graphqlSchema, graphqlDocument, customScalars, transform } = config

  const graphqlSchema_ =
    typeof graphqlSchema === 'string' ? buildSchema(graphqlSchema) : graphqlSchema

  const graphqlDocument_ =
    typeof graphqlDocument === 'string' ? parse(graphqlDocument) : graphqlDocument

  const operations = getBridgeOperations<T>(
    graphqlSchema_,
    graphqlDocument_,
    customScalars,
    transform,
  )

  return {
    getExpressMiddleware: (
      executor: GraphQLExecutor<ThisRequest, ThisResponse>,
      config?: CreateMiddlewareConfig,
    ) => createExpressMiddleware(operations, executor, config),
    getOpenAPISchema: (config: CreateOpenAPISchemaConfig<T>): OpenAPIV3.Document<T> =>
      createOpenAPISchemaWithValidate<T>(operations, config),
  }
}
