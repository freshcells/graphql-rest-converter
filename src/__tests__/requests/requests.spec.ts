import { bridge, gqlSchema } from '../bridgeFixtures'
import express from 'express'
import { createSchemaExecutor } from '../../graphQLExecutor'
import request from 'supertest'

const app = express()

app.use(
  bridge.getExpressMiddleware(createSchemaExecutor(gqlSchema), {
    validateResponse: true,
    validateRequest: true,
  })
)

describe('Requests', () => {
  it('should GET `/samples`', async () => {
    const response = await request(app).get('/samples').expect(200)
    expect(response.body).toMatchSnapshot({
      samples: [
        {
          myId: expect.any(Number),
        },
        {
          myId: expect.any(Number),
        },
      ],
    })
  })
  it('should GET `/sample/id`', async () => {
    const response = await request(app).get('/sample/1').expect(200)
    expect(response.body).toMatchSnapshot({
      getSample: {
        myId: expect.any(Number),
      },
    })
  })

  it('should fail GET `/sample/id`, if id is not a number', async () => {
    const response = await request(app).get('/sample/not-a-number').expect(400)
    expect(response.body).toMatchSnapshot()
  })

  it('should POST `/sample`', async () => {
    const response = await request(app)
      .post('/sample')
      .send({ name: 'Input', moreData: [] })
      .expect(200)
    expect(response.body).toMatchSnapshot({
      createSample: {
        id: expect.any(Number),
      },
    })
  })

  it('should DELETE `/sample`', async () => {
    const response = await request(app).delete('/sample').query({ id: 1 }).expect(200)
    expect(response.body).toMatchSnapshot({
      deleteSample: expect.any(Boolean),
    })
  })

  it('should fail on DELETE `/sample`', async () => {
    const response = await request(app).delete('/sample').query({ id: 'not-an-id' }).expect(400)
    expect(response.body).toMatchSnapshot()
  })

  it('should PUT `/sample`', async () => {
    const response = await request(app)
      .put('/sample/1')
      .send({ name: 'New Input', moreData: [] })
      .expect(200)
    expect(response.body).toMatchSnapshot()
  })

  it('should fail POST `/sample` if invalid body', async () => {
    const response = await request(app)
      .post('/sample')
      .send({ somethingUnknown: 'Input', moreData: [] })
      .expect(400)
    expect(response.body).toMatchSnapshot()
  })

  it('should not handle other verbs then defined', async () => {
    await request(app).post('/samples').expect(404)
    await request(app).delete('/samples').expect(404)
    await request(app).put('/samples').expect(404)
  })
})
