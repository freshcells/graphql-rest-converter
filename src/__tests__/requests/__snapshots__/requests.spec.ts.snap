// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Requests should DELETE \`/sample\` 1`] = `
Object {
  "deleteSample": Any<Boolean>,
}
`;

exports[`Requests should GET \`/sample/id\` 1`] = `
Object {
  "getSample": Object {
    "myId": Any<Number>,
    "myName": "Hello World",
  },
}
`;

exports[`Requests should GET \`/samples\` 1`] = `
Object {
  "samples": Array [
    Object {
      "myId": Any<Number>,
      "myName": "Hello World",
    },
    Object {
      "myId": Any<Number>,
      "myName": "Hello World",
    },
  ],
}
`;

exports[`Requests should POST \`/sample\` 1`] = `
Object {
  "createSample": Object {
    "id": Any<Number>,
    "name": "Hello World",
  },
}
`;

exports[`Requests should PUT \`/sample\` 1`] = `
Object {
  "updateSample": Object {
    "name": "Hello World",
  },
}
`;

exports[`Requests should fail GET \`/sample/id\`, if id is not a number 1`] = `
Object {
  "errors": Array [
    Object {
      "errorCode": "type.openapi.requestValidation",
      "location": "path",
      "message": "must be integer",
      "path": "id",
    },
  ],
  "status": 400,
}
`;

exports[`Requests should fail POST \`/sample\` if invalid body 1`] = `
Object {
  "errors": Array [
    Object {
      "errorCode": "required.openapi.requestValidation",
      "location": "body",
      "message": "must have required property 'name'",
      "path": "name",
    },
  ],
  "status": 400,
}
`;

exports[`Requests should fail on DELETE \`/sample\` 1`] = `
Object {
  "errors": Array [
    Object {
      "errorCode": "type.openapi.requestValidation",
      "location": "query",
      "message": "must be integer",
      "path": "id",
    },
  ],
  "status": 400,
}
`;
